### PR TITLE
Nodemailer 10 support

### PR DESCRIPTION
**Pull request by thoda-dev**
## Motivation
Nodemailer 10 was released in September 2026. Bumping the dependency is not enough this time: `nodemailer` is now written in TypeScript and ships its own type definitions, which take precedence over `@types/nodemailer` and use a different layout than the DefinitelyTyped ones. The transport compiles against the bundled types only after a few adjustments, even though nothing changes at runtime (the whole test suite passes on 10.0.9 without touching the code).

Nodemailer 10 also requires Node.js 20 or newer, which the CI currently does not meet.

## Changes
- **CI**: run the test workflow on Node.js 20.20.2 and align `.tool-versions` (nodemailer 10 declares `engines.node >= 20`, so `yarn install --frozen-lockfile` would fail on Node 18).
- **Dependencies**: `nodemailer` dev dependency bumped to `^10.0.0`; peer range widened to `^9.0.1 || ^10.0.0` so users on 9 are not warned. `@types/nodemailer` is kept as an optional peer for nodemailer 9 users, it is not needed with 10.
- **Type imports**: `nodemailer/lib/mailer` no longer exposes named exports (`Address`, `Attachment`, `AttachmentLike`, `Headers`) nor a CommonJS `export =`. The code now uses a default import and the `Mail.*` namespace, which is the layout nodemailer 10 explicitly preserved for `@types/nodemailer` compatibility. The sources type-check against both nodemailer 10 (bundled types) and nodemailer 9 + `@types/nodemailer` 6.
- **Adapters**: the bundled types are stricter and closer to what nodemailer actually accepts, so the adapters now handle those inputs instead of assuming the loose DefinitelyTyped shapes:
  - recipients: `name`/`address` are optional on address objects, arrays can be nested and address groups (`{ name, group: [...] }`) are expanded; `from` accepts an array like the other address fields (first one is used, as for `reply_to`).
  - headers: a single `{ key, value }` object is accepted, and number/boolean/Date/address/`{ prepared, value }` values are converted to strings; `null`/`undefined` values are skipped.
  - attachments: content resolution is delegated to `adaptContent`, which also accepts a content descriptor object.
  - internal rename `adaptReplyToRecipient` → `adaptFirstRecipient` (shared by `from` and `reply_to`; not part of the public API).
- **Types**: the nodemailer input shapes used by the adapters (`NodemailerAddress`, `NodemailerRecipients`, `NodemailerContent`) live in `src/types/transport.ts` next to `MailtrapMailOptions`.
- **Tests**: 8 new adapter tests covering the cases above (444 total).
- **README**: nodemailer 9 and 10 supported, Node.js 20+ required for 10, `@types/nodemailer` only needed with 9.

## How to test
- [ ] `yarn install --frozen-lockfile && yarn lint && yarn test` on Node.js 20+ (nodemailer 10.0.9 installed): lint, tsc and 444 tests pass.
- [ ] Backward compatibility with the peer range: install `nodemailer@^9` alongside `@types/nodemailer@^6` and run `tsc -p tsconfig.build.json --noEmit` — the sources still type-check (the new adapter tests use nodemailer 10 shapes and are only meant to compile against the dev dependency).
- [ ] `yarn build` and check that `dist/types/transport.d.ts` and `dist/adapters/*.d.ts` only reference `nodemailer/lib/mailer` through the default import.
- [ ] Send a message through `MailtrapTransport` with nodemailer 10 (see `examples/sending/transport.ts`) and confirm the payload is unchanged.

## Summary by CodeRabbit

- **New Features**
  - Added compatibility with Nodemailer 10 while maintaining support for Nodemailer 9.
  - Added support for grouped and nested recipients.
  - Expanded header conversion to handle dates, booleans, numbers, addresses, arrays, and empty values.

- **Bug Fixes**
  - Improved recipient and header conversion for consistent message formatting.
  - Improved validation for missing sender addresses and empty attachment content.

- **Documentation**
  - Updated setup guidance for Nodemailer versions and TypeScript type definitions.

### CODE REVIEW

File: src/adapters/content.ts [hunk 1]
```diff
@@ -9,7 +10,7 @@ import { AttachmentLike } from "nodemailer/lib/mailer";
  * Otherwise reads file.
  */
 export default function adaptContent(
-  content: string | Buffer | Readable | AttachmentLike
+  content: NodemailerContent
 ): string | Buffer {
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/adapters/content.ts:14]: This check predates this PR, only the input type was widened here. Leaving it as is to keep the scope on nodemailer 10 support.

File: src/adapters/recipients.ts [hunk 2]
```diff
@@ -13,52 +30,39 @@ export function adaptSingleRecipient(
     return { email: recipient };
   }
 
-  return { name: recipient.name, email: recipient.address };
+  return {
+    ...(recipient.name !== undefined && { name: recipient.name }),
+    email: recipient.address ?? "",
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/adapters/recipients.ts:35]: Fixed in 8e7b191: `adaptMail` now rejects a `from` without a usable address, covered by the new test cases.

File: src/adapters/headers.ts [hunk 3]
```diff
@@ -1,42 +1,83 @@
...
+  if (typeof value === "number" || typeof value === "boolean") {
+    return String(value);
+  }
+
+  if (value instanceof Date) {
+    return value.toUTCString();
+  }
+
+  if (Array.isArray(value)) {
+    return adaptHeaderValue(value[0]); // TODO: support multiple value per header
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/adapters/headers.ts:31]: Pre-existing behaviour (see the original TODO), not changed by this PR. Keeping the scope on nodemailer 10 support.

File: src/adapters/content.ts [hunk 4]
```diff
@@ -9,7 +10,7 @@ import { AttachmentLike } from "nodemailer/lib/mailer";
  * Otherwise reads file.
  */
 export default function adaptContent(
-  content: string | Buffer | Readable | AttachmentLike
+  content: NodemailerContent
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/adapters/content.ts:13]: Pre-existing behaviour, this PR only widens the input type. Keeping the scope on nodemailer 10 support.

File: src/types/transport.ts [hunk 5]
```diff
@@ -8,6 +9,40 @@ import {
...
+  | NodemailerAddress
+  | NodemailerRecipients[];
+
+/**
+ * Object pointing to a content instead of carrying it.
+ */
+type NodemailerContentObject = {
+  content?: NodemailerContent | undefined;
+  path?: unknown;
+};
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/types/transport.ts:36]: Narrowed `path` to `string | false | Url` in d3feae7. The `readFileSync` fallback itself predates this PR and is left as is.

File: src/adapters/headers.ts [hunk 6]
```diff
@@ -1,42 +1,87 @@
...
+      return adaptHeaderValue(value.value);
     }
 
-    if (Array.isArray(value)) {
-      [acc[key]] = value; // TODO: support multiple value per header
+    if ("address" in value) {
+      const { name, address } = value as Mail.Address;
 
-      return acc;
+      if (!address) {
```
[the remark this answers is not in the document]
Comment (thoda-dev) [reply src/adapters/headers.ts]: Fixed in d3feae7, whitespace-only addresses are now skipped.

File: src/adapters/attachement.ts
Comment (VladimirTaytor) [inline src/adapters/attachement.ts:27]: As we are touching adapter here we probably should do this instead:

```ts
const content = adaptContent(nodemailerAttachment.content);

if (!content) {
  throw new Error(CONTENT_REQUIRED);
}

```

**Review by VladimirTaytor — APPROVED** [review]
Thanks for the input, looks good, suggesting one small improvement to `src/adapters/attachement.ts`

File: src/adapters/attachement.ts
Comment (thoda-dev) [reply src/adapters/attachement.ts:27]: Done in 513692e: the adapted content is now checked and `CONTENT_REQUIRED` is thrown when it is empty, e.g. an empty string or a stream with nothing to read. I kept the initial guard so `adaptContent` still receives a defined value. Also added a test covering both cases.

**Comment by thoda-dev** [comment]
@VladimirTaytor also another thing to check:
I didn't done it in this PR as it's not directly linked but node 20 will be deprecated from Github Action Runner soon (see https://github.blog/changelog/2025-09-19-deprecation-of-node-20-on-github-actions-runners/) I just saw it while pushing on another repo.

I can also upgrade the CI to 24 (I upgraded from 18 to 20 as it's the new requirement of nodemailer 10), or if you prefer, do it on another pr.

and another one: CodeQL v3 will also be deprecated (https://github.blog/changelog/2025-10-28-upcoming-deprecation-of-codeql-action-v3/)

You can see it on your github action log : 
image

Just reply to this comment if you want me to do it.

**Review by izikaj — APPROVED** [review]
Nice job 👍

**Comment by izikaj** [comment]
> also another thing to check

@thoda-dev, any input is valuable and appreciated, especially to make CI use Node 24 😅

**Comment by thoda-dev** [comment]
> > also another thing to check
> 
> @thoda-dev, any input is valuable and appreciated, especially to make CI use Node 24 😅

Ok, I'll work on another PR to update all the CI, when this one will be merge (can create a conflict to the update already done in that one). I prefer to not mix things.

File: src/types/transport.ts [hunk 7]
```diff
@@ -1,5 +1,7 @@
-import NodemailerMail = require("nodemailer/lib/mailer");
+import NodemailerMail from "nodemailer/lib/mailer";
```
Comment (oshchyhol) [inline src/types/transport.ts]: [should] This default import ends up in `dist/types/transport.d.ts`, which every TypeScript user loads through `import { MailtrapTransport } from "mailtrap"`. On nodemailer 9 the types come from `@types/nodemailer`, which uses `export =`, so a project without `esModuleInterop` or `allowSyntheticDefaultImports` can't default-import it. With `skipLibCheck: false` tsc fails inside our d.ts. With `skipLibCheck: true` the error is hidden, but `NodemailerMail.Options` resolves to nothing and a plain `sendMail({ from, to, subject, text })` fails with "'from' does not exist". The same project compiles fine with the current release, so these users break just by upgrading mailtrap, even though the peer range still allows nodemailer 9.

You can import `SendMailOptions` from the package root instead (`import type { SendMailOptions, Transport, Transporter } from "nodemailer"`) and use it in place of `NodemailerMail.Options`. Both `@types/nodemailer` 6 and nodemailer 10 export it there. With that change a nodemailer 9 project compiles with and without those flags.

File: src/adapters/headers.ts [hunk 8]
```diff
@@ -1,42 +1,88 @@
...
+    if ("address" in value) {
+      const { name, address } = value as Mail.Address;
+      const email = address?.trim();
 
-      return acc;
+      if (!email) {
+        return undefined;
+      }
+
+      return name ? `${name} <${email}>` : email;
```
Comment (oshchyhol) [inline src/adapters/headers.ts]: [could] The display name goes in unquoted, so `{ name: "Doe, John", address: "j@x.com" }` becomes `Doe, John <j@x.com>`. A strict RFC 5322 parser reads that as two mailboxes, `Doe` and `John <j@x.com>`. Nodemailer's own transports send `"Doe, John" <j@x.com>` for the same input. Names with `"`, `;`, `<` or non-ASCII characters have the same problem.

Nodemailer leaves a name as it is when it's only word characters and spaces, and otherwise wraps it in quotes and escapes `"` and `\`. Doing the same here would cover the ASCII cases.

File: src/adapters/attachement.ts [hunk 9]
```diff
@@ -24,11 +28,11 @@ export default function adaptAttachment(
...
-  const content =
-    typeof nodemailerAttachment.content === "string" ||
-    nodemailerAttachment.content instanceof Buffer
-      ? nodemailerAttachment.content
-      : nodemailerAttachment.content.read();
+  const content = adaptContent(nodemailerAttachment.content);
+
+  if (!content) {
+    throw new Error(CONTENT_REQUIRED);
+  }
```
Comment (oshchyhol) [inline src/adapters/attachement.ts]: [could] When `adaptAttachment` throws, `sendMail` never finds out. `adaptMail` runs inside nodemailer's `setImmediate` callback with no try/catch, so the throw becomes an uncaught exception and the `sendMail` promise never resolves or rejects. Attaching an empty file does this today: nodemailer hands it over as `content: ''`, and the check above throws. That part isn't new in this PR, but the PR adds a second throw and a test that expects it.

If you wrap this call in a try/catch that passes the error to `callback`, `sendMail` rejects with "Content is required." instead:

https://github.com/mailtrap/mailtrap-nodejs/blob/513692e96edc60b2a138e75f3549adcc576cab2f/src/lib/normalizer.ts#L30-L38

File: .github/workflows/test.yml [hunk 10]
```diff
@@ -9,7 +9,7 @@ jobs:
       - uses: actions/checkout@v2
       - uses: actions/setup-node@v3
         with:
-          node-version: 18.12.1
+          node-version: 20.20.2
```
Comment (oshchyhol) [inline .github/workflows/test.yml]: [could] With CI on Node 20 only, nothing tests the versions `engines` still promises (`>=16.20.1`). On Node 18, `yarn install --frozen-lockfile` now fails on nodemailer 10's engines check, even though the tests themselves pass there with `--ignore-engines`. CI also never runs against nodemailer 9 with `@types/nodemailer`, which is where the type break in `src/types/transport.ts` shows up.

You could add a Node 18 job (with `--ignore-engines`, or with nodemailer 9 installed), or raise `engines.node` and mention it in the release notes.

File: src/adapters/attachement.ts [hunk 11]
```diff
@@ -24,11 +28,11 @@ export default function adaptAttachment(
     throw new Error(CONTENT_REQUIRED);
   }
 
-  const content =
-    typeof nodemailerAttachment.content === "string" ||
-    nodemailerAttachment.content instanceof Buffer
-      ? nodemailerAttachment.content
-      : nodemailerAttachment.content.read();
+  const content = adaptContent(nodemailerAttachment.content);
```
Comment (oshchyhol) [inline src/adapters/attachement.ts]: [should] This lets an attachment read a file from disk even when `disableFileAccess` is on. nodemailer only looks for `path` one level deep, so it passes `{ filename: "x.txt", content: { content: { path: "/app/.env" } } }` through untouched. `adaptContent` then walks down to `path` and calls `readFileSync`, and the file goes out as the attachment. I ran it: with `disableFileAccess: true` the payload carried the file's contents, while `{ path }` and `{ content: { path } }` were rejected with `EFILEACCESS`. On main the same input throws a TypeError and nothing is read, and nodemailer's own transports reject that shape too.

By the time the adapter runs, nodemailer has already resolved every attachment form it supports into a string or a Buffer. Accepting only those here and throwing `CONTENT_REQUIRED` for anything else would close this. With the try/catch from the comment below, `sendMail` would then reject instead of crashing.

File: src/adapters/headers.ts [hunk 12]
```diff
@@ -1,42 +1,88 @@
...
+  }
 
-      return acc;
-    }, {} as MailtrapHeaders);
+  if (typeof value === "string") {
+    return value;
   }
 
-  const headerKeys = Object.keys(nodemailerHeaders);
+  if (typeof value === "number" || typeof value === "boolean") {
```
Comment (oshchyhol) [inline src/adapters/headers.ts]: [could] Header values now go out where nodemailer drops them. `headers: { "X-Flag": false, "X-Retry": 0, "X-Date": new Date(0) }` reaches the API as `"false"`, `"0"` and `"Thu, 01 Jan 1970 00:00:00 GMT"`. nodemailer's own transports send none of these three, and main dropped them too. Empty and whitespace-only strings also go out here while nodemailer drops them, though that part isn't new.

CR/LF isn't removed either. nodemailer folds `"a\r\nInjected: yes"` into one line, while here it goes out as is. The PR adds the same for address names: `{ name: "Eve\r\nBcc: victim@x.com", address: "j@x.com" }` reaches the API with the line break inside. Skipping falsy and blank values and replacing CR/LF with a space would match nodemailer.

File: src/adapters/recipients.ts [hunk 13]
```diff
@@ -1,10 +1,27 @@
...
+ * Address groups (`{ name, group: [...] }`) are expanded into their members.
+ */
+function flattenRecipients(
+  recipients: NodemailerRecipients
+): Array<string | NodemailerAddress> {
+  if (Array.isArray(recipients)) {
+    return recipients.flatMap(flattenRecipients);
+  }
+
+  if (typeof recipients !== "string" && recipients.group) {
```
Comment (oshchyhol) [inline src/adapters/recipients.ts]: [could] Checking `group` before `address` drops the object's own address. `to: { name: "Team", address: "team@x.com", group: [{ address: "a@x.com" }] }` goes out to `a@x.com` only. nodemailer keeps the object as it is when `address` is set, so this does reach the adapter, and both main and nodemailer's own transports send to `team@x.com`. Checking `address` first would match them.

File: src/adapters/recipients.ts (hunk 2 above)
Comment (oshchyhol) [inline src/adapters/recipients.ts]: [could] A recipient without an address goes out as `email: ""`. nodemailer turns `{ name: "Bob" }` into `{ name: "[object Object]", address: "" }`, so `to: [{ name: "Bob" }, "a@x.com"]` reaches the API as `[{ name: "[object Object]", email: "" }, { name: "", email: "a@x.com" }]`. For `replyTo: [{ name: "Support" }, "support@x.com"]` the empty one is sent, since the first entry wins. nodemailer's own transports leave these entries out of the envelope and the headers.

This isn't new in this PR, but the PR adds the check for `from` only. Skipping entries without an address here would cover `to`, `cc`, `bcc` and `replyTo` as well.

File: src/adapters/headers.ts [hunk 14]
```diff
@@ -1,42 +1,88 @@
...
+      return nodemailerHeaders.map(({ key, value }) => [key, value]);
+    }
+
+    // Single `{ key, value }` header, handled the same way nodemailer does in `setHeader`.
+    const { key, value } = nodemailerHeaders as {
+      key?: unknown;
+      value?: unknown;
+    };
+
+    if (typeof key === "string" && "value" in nodemailerHeaders) {
```
Comment (oshchyhol) [inline src/adapters/headers.ts]: [nitpick] This check doesn't match what nodemailer does with a single `{ key, value }` object in custom headers. `{ key: "", value: "x" }` becomes a header with an empty name here, while nodemailer treats the object as plain headers and sends `Value: x`. `{ key: "X-Count", value: 0 }` sends `X-Count: 0` here, and nodemailer sends no `X-Count` header.

File: src/adapters/headers.ts [hunk 15]
```diff
@@ -1,42 +1,88 @@
...
 
-    if (typeof value === "string") {
-      acc[key] = value;
+  if (Array.isArray(value)) {
+    return adaptHeaderValue(value[0]); // TODO: support multiple value per header
+  }
 
-      return acc;
+  if (typeof value === "object") {
+    if ("value" in value) {
```
Comment (oshchyhol) [inline src/adapters/headers.ts]: [nitpick] Input that refers to itself now overflows the stack. A header value like `const o = {}; o.value = o`, or an address object with an `address` whose `group` contains itself, ends in "Maximum call stack size exceeded". It's thrown inside nodemailer's callback, so it crashes the process for the same reason as in the attachment comment. Main and nodemailer's own transports handle both. It takes contrived input, so a depth limit would be enough.

File: src/adapters/mail.ts [hunk 16]
```diff
@@ -21,16 +18,18 @@ const { SUBJECT_REQUIRED, FROM_REQUIRED } = ERRORS;
...
-    from: adaptSingleRecipient(data.from),
+    from,
     to: adaptRecipients(data.to),
     cc: adaptRecipients(data.cc),
     bcc: adaptRecipients(data.bcc),
-    reply_to: adaptReplyToRecipient(data.replyTo),
+    reply_to: adaptFirstRecipient(data.replyTo),
   };
 
   if (data.headers) {
```
Comment (oshchyhol) [inline src/adapters/mail.ts]: [could] `normalize()` has already built `data.normalizedHeaders`, which includes the `references`, `in-reply-to` and `list-*` headers. The adapter only reads `data.headers`, so `references`, `inReplyTo` and `list: { unsubscribe }` never reach the API. That isn't new in this PR, but since the headers code is being rewritten anyway, reading from `normalizedHeaders` would pick them up.

File: src/adapters/headers.ts [hunk 17]
```diff
@@ -1,42 +1,88 @@
...
+      return adaptHeaderValue(value.value);
     }
 
-    if (Array.isArray(value)) {
-      [acc[key]] = value; // TODO: support multiple value per header
+    if ("address" in value) {
+      const { name, address } = value as Mail.Address;
```
Comment (oshchyhol) [inline src/adapters/headers.ts]: [nitpick] This casts to `Mail.Address`, while the PR adds `NodemailerAddress` in `src/types/transport.ts` for the same shape. Using that type here would keep a single definition.

**Review by oshchyhol — COMMENTED** [review]
Thanks for putting this together! Supporting nodemailer 10 while keeping 9 working is fiddly!

I ran the cases in the comments below against the built package, on nodemailer 9 and 10, with current main for comparison. They're tagged by how much they matter:
- [should]: has to be fixed before merging. There are currently two of these: attachments can read local files even with `disableFileAccess` on, and TypeScript projects still on nodemailer 9 stop compiling.
- [could]: improvements, ok to leave for later.
- [nitpick]: minor, take them or leave them :)

File: src/types/transport.ts (hunk 7 above)
Comment (thoda-dev) [reply src/types/transport.ts]: Fixed in 0d42479: `SendMailOptions` from the package root, as suggested. I also dropped the two remaining `nodemailer/lib/mailer` imports in the adapters, since they end up in the same d.ts graph. Checked on a consumer project with nodemailer 9 and 10, with and without `esModuleInterop`.

File: src/adapters/headers.ts (hunk 17 above)
Comment (thoda-dev) [reply src/adapters/headers.ts]: Done in 0d42479: the cast uses `NodemailerAddress` now, the adapter no longer imports `nodemailer/lib/mailer`.

File: src/adapters/attachement.ts (hunk 9 above)
Comment (thoda-dev) [reply src/adapters/attachement.ts]: Fixed in 8d7c26a: caught in `normalizer.ts`, so it covers every adapter throw.

File: src/adapters/attachement.ts (hunk 11 above)
Comment (thoda-dev) [reply src/adapters/attachement.ts]: Fixed in cd095dd: only strings and Buffers are accepted now. Re-ran your cases — the nested shapes reject with "Content is required.", and streams, Buffers, `path` and data URIs still go through since nodemailer resolves them first.

File: src/adapters/headers.ts (hunk 8 above)
Comment (thoda-dev) [reply src/adapters/headers.ts]: Fixed in 33ce202. Non-ASCII names are quoted rather than MIME-encoded, since the API takes UTF-8. nodemailer drops address objects in custom headers altogether — I kept them formatted, say if you'd rather drop them.

File: src/adapters/headers.ts (hunk 12 above)
Comment (thoda-dev) [reply src/adapters/headers.ts]: Fixed in 33ce202: falsy, blank and date values skipped, CR/LF replaced. Only the repeated-header array case still differs from nodemailer, left with the existing TODO.

File: src/adapters/headers.ts (hunk 14 above)
Comment (thoda-dev) [reply src/adapters/headers.ts]: Fixed in d5fe397: read the way `addHeader` does. Two deliberate deviations: a blank name is dropped instead of sent as `: x`, and names keep the case they were given rather than nodemailer's `Key:`/`Value:`.

File: src/adapters/recipients.ts (hunk 13 above)
Comment (thoda-dev) [reply src/adapters/recipients.ts]: Fixed in ab49ea5: the group is only expanded when the object has no address of its own. Matches the envelope on your case and on a mixed array.

File: src/adapters/recipients.ts (hunk 2 above)
Comment (thoda-dev) [reply src/adapters/recipients.ts]: Fixed in bab9b89. Blank names are dropped too, so a plain string no longer goes out as `{ name: "", email }`.

File: src/adapters/headers.ts (hunk 15 above)
Comment (thoda-dev) [reply src/adapters/headers.ts]: Fixed in d1176db: both adapters stop at a nesting limit. The group case doesn't actually reach the adapter (nodemailer mangles it first), but the limit covers it too.

File: src/adapters/mail.ts (hunk 16 above)
Comment (thoda-dev) [reply src/adapters/mail.ts]: Agreed this is a real gap, but I'd rather do it in its own PR.

Reading from `normalizedHeaders` instead of `data.headers` would regress a few things: an array of `{ key, value }` pairs comes back as `{"0":"1","1":"2"}`, a single pair as `{"key":"X-Custom","value":"v"}`, a `Date` is kept as an ISO string where the wire drops it, and CR/LF is left in. So it has to be a merge of just the generated keys (`references`, `in-reply-to`, `list-*`) on top of the adapted headers.

That makes it a feature rather than a fix, with edge cases of its own — `headers: { "List-Unsubscribe" }` together with `list: { unsubscribe }` goes out as two header lines, for one — so it deserves its own tests and changelog entry.

File: .github/workflows/test.yml (hunk 10 above)
Comment (thoda-dev) [reply .github/workflows/test.yml]: The CI updates are going into a separate PR, with the Node 24 bump @izikaj asked for above. A nodemailer 9 + `@types/nodemailer` job is a good call — it's exactly what would have caught the type break — I'll add it there.

On `engines.node` I'd leave it at `>=16.20.1`. It describes what the client needs, and nodemailer is an optional peer: a project on Node 18 can still use the API client, or the transport with nodemailer 9. Node 20 is only required to install nodemailer 10, which constrains contributors and users who opt into it, not the package itself.

**Comment by thoda-dev** [comment]
Thanks for the thorough pass @oshchyhol, that was really useful.

I went through every point — one commit each, with the reasoning in the replies inline. Both [should] ones are fixed, and the [could]/[nitpick] ones too.

I resolved the threads that were settled, and left the three that may deserve a second look unresolved: the address objects in custom headers, `engines.node`, and `normalizedHeaders`. Reply on them if you want to add anything, otherwise feel free to resolve them yourself.

A few extra cases turned up along the way and are handled as well: blank display names no longer go out as `{ name: "", email }`, headers with a blank name are skipped, and the remaining `nodemailer/lib/mailer` imports in the adapters are gone, not just the one in `types/transport.ts`.

Everything was checked against both nodemailer 9 and 10.

File: src/adapters/mail.ts (hunk 16 above)
Comment (oshchyhol) [reply src/adapters/mail.ts]: true, let's skip this. it looked like an easy improvement on the surface

File: package.json [hunk 18]
```diff
@@ -48,7 +48,7 @@
   },
   "peerDependencies": {
     "@types/nodemailer": "^6.4.9",
-    "nodemailer": "^9.0.1"
+    "nodemailer": "^9.0.1 || ^10.0.0"
```
Comment (oshchyhol) [inline package.json]: [should] With nodemailer 10, TypeScript projects using native ESM (`"type": "module"` + `module: nodenext`) lose the Mailtrap `sendMail` typing. Passing `category` fails with `TS2353 'category' does not exist in type 'SendMailOptions'`. Nodemailer 10 ships separate type files for `require` and `import`, and since this package is CommonJS, the `createTransport` augmentation only reaches the `require` copy. Nodemailer 9 with `@types/nodemailer` and CommonJS consumers are fine.

https://github.com/mailtrap/mailtrap-nodejs/blob/d1176dbb075f673479871160fd22b6ebcea8c5d4/src/lib/transport.ts#L57-L62

A fix that worked for me: repeat the augmentation in an ESM file and reference it from `src/lib/transport.ts`.

```ts
// src/lib/transport-esm.mts
import type { MailtrapTransporter } from "../types/transport.js";
import type { MailtrapTransportInstance } from "./transport.js";

declare module "nodemailer" {
  export function createTransport(
    transport: MailtrapTransportInstance
  ): MailtrapTransporter;
}
```

It also needs `/// <reference path="./transport-esm.mts" preserve="true" />` at the top of `src/lib/transport.ts`, `export type { MailtrapTransport as MailtrapTransportInstance };` in the same file, and `rimraf dist/lib/transport-esm.mjs` in the build script to drop an empty `.mjs` that tsc emits. I checked `node10`, `nodenext` CommonJS/ESM and `bundler` consumers against the packed build. All pass, and lint and tests are green.

**Review by oshchyhol — APPROVED** [review]

File: package.json (hunk 18 above)
Comment (thoda-dev) [reply package.json]: Fixed in 1a803cc, applied as suggested.

I checked the packed build against `node10`, `nodenext` CJS/ESM and `bundler` consumers on both nodemailer 9 and 10 — `nodenext` ESM on nodemailer 10 was the only failing one, and all pass now.

**Comment by thoda-dev** [comment]
@oshchyhol I did a full review pass over the branch before calling it done, and a few things came up. Fixed in d4d576f.

The `disableFileAccess` bypass from the attachment thread has a twin on `text` / `html`.
nodemailer filters two levels deep there instead of one, so `text: { content: { content: { path: "/app/.env" } } }` still reached `adaptContent`, which read the file and sent its contents. It's on `main` too, but it's the same hole, so it's closed the same way: `adaptContent` now takes a string or a Buffer and throws `CONTENT_REQUIRED` for anything else.

That made the inline check in `adaptAttachment` a duplicate, so it calls `adaptContent`
again as @VladimirTaytor originally suggested — safe now that the function reads nothing.
A self-referencing `text` also stops overflowing the stack.

The last one is smaller: header names are trimmed when written, not only when checked,
so `{ "  X-Foo  ": "v" }` goes out as `X-Foo`, the way nodemailer writes it.

I also ran the adapters from `main` and from the branch over the same normalized messages
on both nodemailer 9 and 10: same 12 intentional differences as before these changes,
nothing new.
